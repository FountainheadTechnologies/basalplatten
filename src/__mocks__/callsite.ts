module.exports = () => ([{
}, {
  getFileName: () => 'test-caller.js'
}]);
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
