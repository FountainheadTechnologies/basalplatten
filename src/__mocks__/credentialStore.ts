export const credentialStore = {
  setToken: jest.fn(),
  getToken: jest.fn(),
  clearToken: jest.fn()
}
