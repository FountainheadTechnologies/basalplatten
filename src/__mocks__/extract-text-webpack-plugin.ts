module.exports = class ExtractTextPlugin {
  extract() {}
};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
