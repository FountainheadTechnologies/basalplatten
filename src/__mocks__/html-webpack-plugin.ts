module.exports = class HtmlWebpackPlugin {};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
