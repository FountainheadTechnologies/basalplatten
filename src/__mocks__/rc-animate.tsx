import React from 'react';

const RcAnimate = props => (
  <div>{props.component}</div>
)

RcAnimate.displayName = 'RcAnimate';

export default RcAnimate;
