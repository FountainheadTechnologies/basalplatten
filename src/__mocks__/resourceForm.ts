const submit = jest.fn();

export const createToken = {
  schema: {
    required: ['username', 'password'],
    properties: {
      username: { type: 'string' },
      password: { type: 'string' }
    },
  },

  submit
}

export const createPerson = {
  schema: {
    required: ['forename', 'surname'],
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  },

  submit
}

export const updatePerson = {
  schema: {
    required: [],
    default: {
      forename: 'Test',
      surname: 'Person',
      date_of_birth: '1985-08-02T23:00:00.000Z',
      credits: 33,
      favourite_colour: 'blue'
    },
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      credits: { type: 'number' },
      enabled: { type: 'boolean' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      },
      favourite_colour: {
        enum: ['red', 'green', 'blue']
      }
    }
  }
}
