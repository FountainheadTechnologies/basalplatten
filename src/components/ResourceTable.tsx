import React from 'react';
import Table from 'antd/lib/table';
import { TableProps } from 'antd/lib/table/Table';
import { ColumnProps } from 'antd/lib/table/Column';
import { PaginationProps } from 'antd/lib/pagination';
import { Resource } from '@optics/hal-client';
import { equals } from 'ramda';
import 'antd/lib/table/style';

import { filtersToWhereParam, sorterToOrderParam, Sorter, SORT_TO_ORDER } from '../hoc/stateParamsControlledTable';

export interface Props extends TableProps<any> {
  resource: Resource;
  rel: string;
  name?: string;
  embedded?: string;
}

export interface State {
  loading: boolean;
  total: number;
  rows: Resource[];
  params: {};
}

export const columnsToWhereParam = (columns: ColumnProps<any>[] | undefined) => {
  if (!columns) {
    return;
  }

  const result: { [key: string]: string | string[] } = {};

  columns.forEach(column => {
    const key = column.key || column.dataIndex;

    if (!key) {
      return;
    }

    const values = column.filteredValue;

    if (!values || !values.length) {
      return;
    }

    result[key] = column.filterMultiple ? values : values[0];
  });

  return Object.keys(result).length ? result : undefined;
}

export const columnsToOrderParam = (columns: ColumnProps<any>[] | undefined) => {
  if (!columns) {
    return;
  }

  const result: { [key: string]: string } = {};

  columns.forEach(column => {
    const key = column.key || column.dataIndex;

    if (!key) {
      return;
    }

    if (typeof column.sortOrder !== 'string') {
      return;
    }

    result[key] = SORT_TO_ORDER[column.sortOrder];
  });

  return Object.keys(result).length ? result : undefined;
}

export const propsToParams = (props: Props) => ({
  page: (typeof props.pagination === 'object') ? props.pagination.current : undefined,
  where: columnsToWhereParam(props.columns),
  order: columnsToOrderParam(props.columns)
})

export class ResourceTable extends React.PureComponent<Props, State> {
  state: State = {
    loading: false,
    total: 0,
    rows: [],
    params: propsToParams(this.props)
  }

  render() {
    const {
      resource,
      rel,
      name,
      embedded,
      ...tableProps
    } = this.props;

    const { loading, rows } = this.state;

    return (
      <Table
        {...tableProps}
        pagination={this.pagination()}
        loading={loading}
        dataSource={rows}
        onChange={this.onChange}
      />
    );
  }

  componentDidMount() {
    this.fetch();
  }

  componentWillReceiveProps(nextProps: Props) {
    const params = this._nextParams(nextProps);

    if (params) {
      this.setState({ params });
      this.fetch(params);
    }
  }

  pagination = () => {
    const { pagination } = this.props;
    const { total } = this.state;

    if (pagination === false) {
      return false;
    }

    return {
      ...pagination as {},
      total
    }
  }

  fetch = (params = this.state.params) => {
    const { resource, rel, name, embedded } = this.props;

    this.setState({ loading: true });

    const link = name ?
      resource.linkNamed(rel, name) : resource.link(rel);

    return link.fetch(params)
      .then((result: Resource) => {
        this.setState({
          total: result.properties.count,
          rows: result.embedded(embedded || rel) as Resource[]
        })
      })
      .finally(() => {
        this.setState({ loading: false });
      });
  }

  onChange = (pagination: PaginationProps, filters: any, sorter: Sorter) => {
    const { columns, onChange } = this.props;

    const params = {
      page: pagination.current || 1,
      where: filtersToWhereParam(columns, filters),
      order: sorterToOrderParam(sorter)
    };

    this.setState({ params });
    this.fetch(params);

    if (onChange) {
      onChange(pagination, filters, sorter);
    }
  }

  protected _nextParams = (nextProps: Props) => {
    if (equals(nextProps.columns, this.props.columns)) {
      return;
    }

    const params = propsToParams(nextProps);

    if (equals(params, this.state.params)) {
      return;
    }

    return params;
  }
}
