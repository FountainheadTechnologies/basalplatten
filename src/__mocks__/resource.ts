import { createToken } from './resourceForm';

export const _deferred = {
  resolve: undefined,
  reject: undefined
};

export const fetch = jest.fn((binding = {}, params = {}) => {
  return new Promise((resolve, reject) => {
    Object.assign(_deferred, { resolve, reject });
  });
});

export const api = {
  properties: {},
  meta: {},

  config: {
    headers: {}
  },

  link: (rel: string) => ({
    fetch: (params = {}) => fetch({ rel }, params),
  }),

  linkNamed: (rel: string, name: string) => ({
    fetch: (params = {}) => fetch({ rel, name }, params)
  }),

  formNamed(rel: string, name: string) {
    if (rel === 'token' && name === 'create') {
      return createToken;
    }
  }
}
