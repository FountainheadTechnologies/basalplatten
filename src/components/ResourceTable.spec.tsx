import React from 'react';
import promiseFinally from 'promise.prototype.finally';
import { mount } from 'enzyme';

import { ResourceTable } from './ResourceTable';
import { api, fetch, _deferred } from '../__mocks__/resource';

promiseFinally.shim();

beforeEach(() => fetch.mockClear());

describe('default behaviour', () => {
  let root;

  beforeEach(() => {
    root = mount((
      <ResourceTable
        resource={api}
        rel="widgets"
      />
    ));
  })

  it('renders an Ant Table', () => {
    expect(root).toMatchSnapshot();
  });

  it('fetches data on mount', () => {
    expect(fetch).toHaveBeenCalledWith({ rel: 'widgets' }, {});
  });

  describe('fetch', () => {
    const widgets = [{
      properties: {
        name: 'Widget #1',
      }
    }, {
      properties: {
        name: 'Widget #2'
      }
    }];

    const result = {
      embedded: jest.fn((name: string) => widgets),
      properties: {
        total: widgets.length
      }
    }

    it('sets Table props before and after loading', () => {
      const table = root.find('Table');
      const promise = root.instance().fetch();

      expect(table.prop('loading')).toBe(true);
      expect(table.prop('pagination')).toEqual({ total: 0 });
      expect(table.prop('dataSource')).toEqual([]);
      _deferred.resolve(result);

      return promise.then(() => {
        expect(table.prop('loading')).toBe(false);
        expect(table.prop('pagination')).toEqual({ total: 2 });
        expect(table.prop('dataSource')).toBe(widgets);
      });
    });
  })

  describe('when table params change', () => {
    beforeEach(() => {
      const pagination = { current: 3 };
      const filters = {};
      const sorter = { columnKey: 'name', order: 'ascend' };

      root.find('Table').props().onChange(pagination, filters, sorter);
    });

    it('re-fetches data with new params', () => {
      expect(fetch).toHaveBeenCalledWith({ rel: 'widgets' }, {
        page: 3,
        order: {
          name: 'asc'
        }
      });
    });

    it('re-uses params on additional calls to .fetch()', () => {
      fetch.mockClear();
      root.instance().fetch();

      expect(fetch).toHaveBeenCalledWith({ rel: 'widgets' }, {
        page: 3,
        order: {
          name: 'asc'
        }
      });
    });
  });
});

describe('when `name` is specified', () => {
  let root;

  beforeEach(() => {
    root = mount((
      <ResourceTable
        resource={api}
        rel="widgets"
        name="collection"
      />
    ));
  })

  it('fetches link with `rel` and `name`', () => {
    expect(fetch).toHaveBeenCalledWith({ rel: 'widgets', name: 'collection' }, {});
  });
});
