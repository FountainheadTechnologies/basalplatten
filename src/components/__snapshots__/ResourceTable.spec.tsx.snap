// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders an Ant Table 1`] = `
<ResourceTable
  rel="widgets"
  resource={
    Object {
      "config": Object {
        "headers": Object {},
      },
      "formNamed": [Function],
      "link": [Function],
      "linkNamed": [Function],
      "meta": Object {},
      "properties": Object {},
    }
  }
>
  <Table
    bordered={false}
    className=""
    dataSource={Array []}
    indentSize={20}
    loading={true}
    locale={Object {}}
    onChange={[Function]}
    pagination={
      Object {
        "total": 0,
      }
    }
    prefixCls="ant-table"
    rowKey="key"
    rowSelection={null}
    showHeader={true}
    size="large"
    useFixedHeader={false}
  >
    <div
      className="ant-table-wrapper"
    >
      <Spin
        className="ant-table-without-pagination ant-table-spin-holder"
        prefixCls="ant-spin"
        size="default"
        spinning={true}
        wrapperClassName=""
      >
        <RcAnimate
          className="ant-spin-nested-loading"
          component="div"
          style={null}
          transitionName="fade"
        >
          <div>
            div
          </div>
        </RcAnimate>
      </Spin>
    </div>
  </Table>
</ResourceTable>
`;
